### PR TITLE
Improve isValidSpreadType check

### LINKED ISSUE #46976: `(T | undefined) & T` has different behavior when T is known versus when T is a type variable
# Bug Report

### 🔎 Search Terms

* or undefined
* and undefined
* undefined generic disjunction
* generic or undefined

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.2#code/LAKA9GAEDKCGC2AHANgUwM6QPYDNICNUALWANwEssAnbAO0gGtasB3egFwE9ENQJIAKkVSRUpWMgCusdqgAmkLj0jlMElrE6ZUAR2nJFWRcJW1Ek9ou6pQSkQIzsAjJAC8kABQuAPpEm05VBxyWnkASkgAMkgnAG5ba0FHACY3T3R2KhCAc0hff0Dg0LkI6Iys2mz4kDskjIBmNI8Ab0hYAC4YyABfPL8AoJDwqMhWjq7u6tqHDIAWJv8mVnp8gaLh6MXmNinEmfYAVibQ0lQaVcKhkpGTs+q+KABZEPJ4CVEADwQUEVwVdHQkhsIBw-gAxuxKPQcAAeAQAPg8EWaoEgaKsygAmk0BH0CoNiqVBNVuqAHpBnrRXu9UF8kGhsHgWMIOERVJAwVhJMgFGDYJJ0CJYJBEFQsPg0PBQKDaBCoZBsk44Z9ZAFMM1ugAaSAAQRVqDVo0gAC9Oh5cRcCRtBD1EeMdcjUejObQMkbjT00rB7iB0ZAqKh2JIqPQUb6-ej+Lj4ALLLTVQoNZB0EZyJZ0EQuTyCCJ0KLULA5LAJcCI+iAHSV41OtGTUCkkDkgCSeHYJlqfw97IAVrGbSEMgWFH9zXi1lciQJtWnICxqAxMOtpeDIVh6Nlksr4wa5Oqtbr9YbWqabd07Z0HaMaxy127j573N6yeG0QGgyGry+I5Xy9Wv3WQG6IA)

### 💻 Code

```ts
// Samples of behavior on known types
// The evaluated type is always equal to the input type
type Test1 = (1 | undefined) & 1; // 1
type Test2 = (string | undefined) & string; // string
type Test3 = ({ a: 1 } | undefined) & { a: 1 }; // { a: 1 } & { a: 1 }
type Test4 = (unknown | undefined) & unknown; // unknown
type Test5 = (never | undefined) & never; // never

// Minimal example of issue
function f<T>() {
    type Y = (T | undefined) & T; // (T | undefined & T)
}

// Minimal example of when this could cause a problem
function g1<T extends {}, A extends { z: (T | undefined) & T }>(a: A) {
    const { z } = a;

    return {
        // T must extend {} so it should be spreadable
        // ERROR: Spread types may only be created from object types. (2698)
        ...z
    };
}

// If the type of z is just T instead of (T | undefined) & T, it works fine
function g2<T extends {}, A extends { z: T }>(a: A) {
    const { z } = a;

    return {
        ...z
    };
}
```

### 🙁 Actual behavior

The type expression `(T | undefined) & T` evaluates to itself when `T` is a type variable.

### 🙂 Expected behavior

Either `(T | undefined) & T` should evaluate to `T` when `T` is a type variable since that's the behavior with known types, or `(T | undefined) & T` should not evaluate to `T` when `T` is a known type (and should instead evaluate to something else in both cases-- `never` maybe?).

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #46976.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -27265,15 +27265,9 @@ namespace ts {
...
-            if (type.flags & TypeFlags.Instantiable) {
-                const constraint = getBaseConstraintOfType(type);
-                if (constraint !== undefined) {
-                    return isValidSpreadType(constraint);
-                }
-            }
-            return !!(type.flags & (TypeFlags.Any | TypeFlags.NonPrimitive | TypeFlags.Object | TypeFlags.InstantiableNonPrimitive) ||
-                getFalsyFlags(type) & TypeFlags.DefinitelyFalsy && isValidSpreadType(removeDefinitelyFalsyTypes(type)) ||
-                type.flags & TypeFlags.UnionOrIntersection && every((type as UnionOrIntersectionType).types, isValidSpreadType));
+            const t = removeDefinitelyFalsyTypes(mapType(type, getBaseConstraintOrType));
```
Comment (sandersn) [inline src/compiler/checker.ts:27268]: love that we have functions named `getBaseConstraintOrType` and `getBaseConstraintOfType`

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -20857,7 +20857,7 @@ namespace ts {
         // flags for the string, number, boolean, "", 0, false, void, undefined, or null types respectively. Returns
         // no flags for all other types (including non-falsy literal types).
         function getFalsyFlags(type: Type): TypeFlags {
-            return type.flags & TypeFlags.Union ? getFalsyFlagsOfTypes((type as UnionType).types) :
+            return type.flags & TypeFlags.UnionOrIntersection ? getFalsyFlagsOfTypes((type as UnionType).types) :
```
Comment (sandersn) [inline src/compiler/checker.ts:20860]: I'm a little surprised that adding intersections to getFalsyFlags didn't change any existing baselines.

**Review by sandersn — APPROVED** [review]

File: src/compiler/checker.ts (hunk 2 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts:20860]: Well, there are precious few intersections that would be affected, pretty much only those like `T & undefined` that will now appear to be falsy.

**Comment by RyanCavanaugh** [comment]
Regression: https://github.com/microsoft/TypeScript/issues/47028#issuecomment-988284040

**Pull request merged**